### PR TITLE
ARROW-8238: [C++] Fix FieldPath type definition

### PR DESCRIPTION
**Pull request by cyb70289**
FieldPath inherits from std::vector and causes link error.
Refactor by replacing inheritance with composition.

### CODE REVIEW

**Comment by cyb70289** [comment]
@bkietz @pitrou @wesm

File: cpp/src/arrow/type.h [hunk 1]
```diff
@@ -1401,20 +1401,30 @@ class ARROW_EXPORT DictionaryUnifier {
...
+  explicit operator bool() const { return !indices_.empty(); }
+  bool operator==(const FieldPath& other) const {
+    return vector() == other.vector();
+  }
+  bool operator!=(const FieldPath& other) const {
+    return !(*this == other);
+  }
+
+  std::vector<int>& vector() { return indices_; }
+  const std::vector<int>& vector() const { return indices_; }
```
Comment (wesm) [inline cpp/src/arrow/type.h]: Nit: I think `indices()` would be more clear

**Review by wesm — COMMENTED** [review]
Looks good aside from the naming issue of the indices accessor

**Comment by wesm** [comment]
```
FAILED: src/arrow/dataset/CMakeFiles/arrow_dataset_objlib.dir/projector.cc.o 
/opt/conda/bin/ccache /opt/conda/bin/x86_64-conda_cos6-linux-gnu-c++  -DARROW_DS_EXPORTING -DARROW_EXTRA_ERROR_CONTEXT -DARROW_HDFS -DARROW_JEMALLOC -DARROW_JEMALLOC_INCLUDE_DIR="" -DARROW_S3 -DARROW_USE_SIMD -DARROW_WITH_BROTLI -DARROW_WITH_LZ4 -DARROW_WITH_SNAPPY -DARROW_WITH_TIMING_TESTS -DARROW_WITH_ZLIB -DARROW_WITH_ZSTD -DURI_STATIC_BUILD -Isrc -I../src -I../src/generated -I../src/parquet -isystem ../thirdparty/flatbuffers/include -isystem jemalloc_ep-prefix/src -isystem ../thirdparty/hadoop/include -Wno-noexcept-type -fvisibility-inlines-hidden -std=c++17 -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /opt/conda/include -fdiagnostics-color=always -ggdb -O0  -Wall -Wno-conversion -Wno-sign-conversion -Wno-unused-variable -Werror -msse4.2  -g -fPIC   -std=gnu++11 -MD -MT src/arrow/dataset/CMakeFiles/arrow_dataset_objlib.dir/projector.cc.o -MF src/arrow/dataset/CMakeFiles/arrow_dataset_objlib.dir/projector.cc.o.d -o src/arrow/dataset/CMakeFiles/arrow_dataset_objlib.dir/projector.cc.o -c ../src/arrow/dataset/projector.cc
../src/arrow/dataset/projector.cc: In member function 'arrow::Status arrow::dataset::RecordBatchProjector::SetDefaultValue(arrow::FieldRef, std::shared_ptr<arrow::Scalar>)':
../src/arrow/dataset/projector.cc:51:21: error: no match for 'operator[]' (operand types are 'arrow::FieldPath' and 'int')
   auto index = match[0];
                     ^
../src/arrow/dataset/projector.cc: In member function 'arrow::Status arrow::dataset::RecordBatchProjector::SetInputSchema(std::shared_ptr<arrow::Schema>, arrow::MemoryPool*)':
../src/arrow/dataset/projector.cc:104:38: error: no match for 'operator[]' (operand types are '__gnu_cxx::__alloc_traits<std::allocator<arrow::FieldPath> >::value_type {aka arrow::FieldPath}' and 'int')
       int matching_index = matches[0][0];
                                      ^
```

Looks like `operator[]` is also needed. Try building with `-DARROW_DATASET=ON` to hit this path

File: cpp/src/arrow/type.h (hunk 1 above)
Comment (cyb70289) [reply cpp/src/arrow/type.h]: will do

**Review by pitrou — APPROVED** [review]
+1.

**Comment by pitrou** [comment]
I rebased on master to clean some CI jobs.

**Comment by pitrou** [comment]
I think we can merge this as is to unblock contributors, then add back some sugar if desired (such as `operator[]`).

**Comment by pitrou** [comment]
Thank you @cyb70289 !

**Pull request closed without merging**